### PR TITLE
Updates for Sortformer training

### PR DESCRIPTION
**Pull request by ipmedenn**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

This PR extends Sortformer training with the functionality required for offline and streaming 8-speaker diarization. It adds new training configurations, subsegment-based data sampling, logits-based and auxiliary losses, block-causal tail augmentation, speaker-counting metrics, and improvements to checkpoint handling and long-recording inference.

The existing probability-based `BCELoss` remains supported, and causal-tail augmentation is disabled by default.

**Collection**: ASR, Speaker Tasks, Common

# Changelog

- Add first-stage offline and second-stage streaming 8-speaker Sortformer training configurations with 80 ms and 10 ms output frames, respectively.
- Add deterministic initialization and sampling based on the configured training seed.
- Add the `init_from_nemo_model` configuration option for warm-starting selected model submodules from a `.nemo` checkpoint.
- Add checkpoint averaging after training and repair scheduler `max_steps` when resuming.
- Update the end-to-end diarization dataset with:
  - a five-field output containing audio, lengths, targets, and speaker identities;
  - subsegment-based training using one or two source chunks;
  - speaker-count-aware chunk sampling;
  - full-duration negative samples when no suitable speech subsegment is available;
  - optional manifest path validation.
- Add an oversampling distributed sampler that produces a fixed number of samples per rank and supports resuming within an epoch.
- Add `BCEWithLogitsLoss` and clean up the existing probability-based `BCELoss`.
- Allow Sortformer forward methods to return speaker and activity logits when required for training while keeping probabilities as the default output.
- Add optional activity and phantom-speaker auxiliary losses.
- Preserve warm-start compatibility with checkpoints that do not contain the optional activity head.
- Add training and validation metrics for speaker-count accuracy and mean absolute error.
- Add training-only block-causal tail augmentation to `TransformerEncoder`: the prefix remains bidirectional, while each tail block attends to the prefix and preceding blocks but not future blocks. The augmentation is disabled by default, and validation and inference continue to use full attention.
- Optimize `max_batch_dur` feature extraction by processing duration-bounded sub-batches without repeated tensor concatenation, and warn when a single recording exceeds the limit.
- Add loading of registered Sortformer models from Hugging Face to `e2e_diarize_speech.py`.
- Add and update unit tests for the dataset, sampler, losses, model forward paths, auxiliary losses, attention masks, checkpoint averaging, metrics, and feature extraction.

# Usage

The offline configuration is intended for the first training stage:

```bash
cd examples/speaker_tasks/diarization/neural_diarizer

python ./sortformer_diar_train.py \
    --config-path='../conf/neural_diarizer' \
    --config-name='sortformer_offline_8spk.yaml' \
    model.train_ds.manifest_filepath=/path/to/train_manifest.json \
    model.validation_ds.manifest_filepath=/path/to/validation_manifest.json
```

The streaming configuration is intended for the second training stage. The first-stage model can be specified through `init_from_nemo_model` in the configuration:

```yaml
init_from_nemo_model:
  stage1:
    path: /path/to/SortformerDiarizer_offline_8spk.nemo
    include: ["encoder", "sortformer_modules"]
```

```bash
python ./streaming_sortformer_diar_train.py \
    --config-path='../conf/neural_diarizer' \
    --config-name='sortformer_streaming_8spk.yaml' \
    model.train_ds.manifest_filepath=/path/to/train_manifest.json \
    model.validation_ds.manifest_filepath=/path/to/validation_manifest.json
```

`e2e_diarize_speech.py` can load either a local checkpoint through `model_path` or a registered Hugging Face model through `pretrained_name`:

```bash
python ./e2e_diarize_speech.py \
    pretrained_name='nvidia/diar_streaming_sortformer_4spk-v2.1' \
    batch_size=1 \
    dataset_manifest=/path/to/diarization_manifest.json
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

Trusted PRs run automatically through copy-pr-bot. For an untrusted PR, a maintainer can trigger CI by commenting
`/ok to test <head-sha>`; repeat this after a new push if the PR remains untrusted.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Speech/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA-NeMo/Speech/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Follow-up to #16032 and #16070.
- User-facing documentation for the new training workflow will be added in a follow-up PR.

### CODE REVIEW

File: examples/speaker_tasks/diarization/conf/neural_diarizer/sortformer_offline_8spk.yaml [hunk 1]
```diff
@@ -0,0 +1,190 @@
+# Offline Sortformer diarizer with a FlexAttention Transformer backbone.
+# This is a first stage training configuration for the Nemotron-3-Diarization model.
```
Comment (nithinraok) [inline examples/speaker_tasks/diarization/conf/neural_diarizer/sortformer_offline_8spk.yaml:2]: Do we want to mention model names before release?

File: examples/speaker_tasks/diarization/conf/neural_diarizer/sortformer_offline_8spk.yaml [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    freq_masks: 2
+    time_masks: 0
+    freq_width: 24
+    time_width: 0.02
+    mask_value: -15  # Fill masked bins near the unnormalized log-mel floor.
+
+  sortformer_modules:
+    _target_: nemo.collections.asr.modules.sortformer_modules.SortformerModules
+    num_spks: ${model.max_num_of_spks}  # Number of output speaker channels.
+    dropout_rate: 0.5  # Sortformer-module dropout probability.
```
Comment (nithinraok) [inline examples/speaker_tasks/diarization/conf/neural_diarizer/sortformer_offline_8spk.yaml:117]: isn;t this quite high?

File: examples/speaker_tasks/diarization/neural_diarizer/checkpoint_averaging.py
Comment (nithinraok) [inline examples/speaker_tasks/diarization/neural_diarizer/checkpoint_averaging.py:1]: move this to `scripts/speaker_tasks/` folder

File: examples/speaker_tasks/diarization/neural_diarizer/checkpoint_averaging.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
```
Comment (nithinraok) [inline examples/speaker_tasks/diarization/neural_diarizer/checkpoint_averaging.py:1]: `SPDX-FileCopyrightText` I don;t think we used this specific text in our previous license header. Could you double check pls

File: examples/speaker_tasks/diarization/neural_diarizer/e2e_diarize_speech.py [hunk 4]
```diff
@@ -36,6 +36,11 @@
     batch_size=1 \
     dataset_manifest=/path/to/diarization_manifest.json
 
+python $BASEPATH/neural_diarizer/e2e_diarize_speech.py \
+    pretrained_name=nvidia/diar_sortformer_4spk-v1 \
```
Comment (nithinraok) [inline examples/speaker_tasks/diarization/neural_diarizer/e2e_diarize_speech.py]: Can we point to recent best model (nemotron-diarization..)

**Review by nithinraok — CHANGES_REQUESTED** [review]
What about doc updates?

File: examples/speaker_tasks/diarization/conf/neural_diarizer/sortformer_offline_8spk.yaml (hunk 2 above)
Comment (ipmedenn) [reply examples/speaker_tasks/diarization/conf/neural_diarizer/sortformer_offline_8spk.yaml:117]: It is high but intentional. It matches the existing offline and streaming 4-speaker Sortformer configs, as well as the default in `SortformerModules`. This value applies to the Sortformer module; the Transformer encoder dropout remains 0.1. I would prefer to keep it consistent with the existing recipes.

File: examples/speaker_tasks/diarization/neural_diarizer/checkpoint_averaging.py (hunk 3 above)
Comment (ipmedenn) [reply examples/speaker_tasks/diarization/neural_diarizer/checkpoint_averaging.py:1]: I double-checked this. The SPDX form is the current repository convention introduced by #16128, which added `SPDX-FileCopyrightText` and `SPDX-License-Identifier: Apache-2.0` to all in-scope source headers. The header in this new file matches the current `main` branch, so I believe it should remain unchanged.

File: examples/speaker_tasks/diarization/neural_diarizer/e2e_diarize_speech.py (hunk 4 above)
Comment (ipmedenn) [reply examples/speaker_tasks/diarization/neural_diarizer/e2e_diarize_speech.py]: The final `nvidia/Nemotron-3-Diarization` model is not published yet, while the currently available `nvidia/Nemotron-3-Diarization-preview` is gated and temporary. Would you prefer that we keep `nvidia/diar_streaming_sortformer_4spk-v2.1` as the stable public example for now, or reference the gated preview and update it after the final release?

File: examples/speaker_tasks/diarization/conf/neural_diarizer/sortformer_offline_8spk.yaml (hunk 1 above)
Comment (ipmedenn) [reply examples/speaker_tasks/diarization/conf/neural_diarizer/sortformer_offline_8spk.yaml:2]: Why not? The model family name is already public through the gated `nvidia/Nemotron-3-Diarization-preview` release.

File: examples/speaker_tasks/diarization/neural_diarizer/checkpoint_averaging.py
Comment (ipmedenn) [reply examples/speaker_tasks/diarization/neural_diarizer/checkpoint_averaging.py:1]: This module is not a standalone utility script: both Sortformer training entry points import it and call it with the live `Trainer` and model immediately after `trainer.fit()`.

The existing files under `scripts/speaker_tasks/` are standalone CLIs, and that directory is not installed as a Python package. Moving this helper there would introduce an `examples` → `scripts` runtime dependency and make the training examples depend on the repository layout or `PYTHONPATH`.

That is why I kept the helper next to the two training entry points. Would you be okay with retaining it here? If you would prefer a packaged shared location, we could instead place it under the `nemo` namespace, but I would avoid importing it from `scripts/`.

File: nemo/collections/asr/modules/sortformer_modules.py [hunk 5]
```diff
@@ -321,12 +349,20 @@ def forward_speaker_sigmoids(self, hidden_out):
...
         """
-        hidden_out = self.dropout(F.relu(hidden_out))
-        hidden_out = self.first_hidden_to_hidden(hidden_out)
-        hidden_out = self.dropout(F.relu(hidden_out))
-        spk_preds = self.single_hidden_to_spks(hidden_out)
-        preds = torch.sigmoid(spk_preds)
-        return preds
+        return torch.sigmoid(self.forward_speaker_logits(hidden_out))
+
+    def forward_activity_logits(self, hidden_out: torch.Tensor) -> Optional[torch.Tensor]:
```
Comment (tango4j) [inline nemo/collections/asr/modules/sortformer_modules.py:354]: Add more description on the purpose of this function as one of the loss calculation heads

File: examples/speaker_tasks/diarization/neural_diarizer/e2e_diarize_speech.py (hunk 4 above)
Comment (tango4j) [reply examples/speaker_tasks/diarization/neural_diarizer/e2e_diarize_speech.py]: I think we can update it to `nvidia/diar_streaming_sortformer_4spk-v2.1` for now..? We should update everything once we do GA on N3D

File: nemo/collections/asr/models/sortformer_diar_models.py [hunk 6]
```diff
@@ -1113,7 +1573,176 @@ def _align_predictions_and_targets(preds, targets, target_lens):
         target_lens = target_lens.clamp(max=common_num_frames)
         return preds, targets, target_lens
 
-    def _get_aux_train_evaluations(self, preds, targets, target_lens) -> dict:
+    @staticmethod
+    def _speaker_count_metrics(
```
Comment (tango4j) [inline nemo/collections/asr/models/sortformer_diar_models.py]: Please consider moving this to a stand-alone file in `metrics` folder.

**Review by tango4j — CHANGES_REQUESTED** [review]
Requesting changes to approve

File: nemo/collections/asr/models/sortformer_diar_models.py (hunk 6 above)
Comment (ipmedenn) [reply nemo/collections/asr/models/sortformer_diar_models.py]: Addressed in d2ee5e09

File: nemo/collections/asr/modules/sortformer_modules.py (hunk 5 above)
Comment (ipmedenn) [reply nemo/collections/asr/modules/sortformer_modules.py:354]: Addressed in d2ee5e09

File: nemo/collections/asr/losses/aux_diarization_loss.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import torch.nn.functional as F
+
+
+def activity_loss(
```
Comment (tango4j) [inline nemo/collections/asr/losses/aux_diarization_loss.py]: This activity_loss should be a class method just like other loss *.py file.
This is because "loss" is always expected to be a class type that has type definitions and variable initializations.

Please follow the convention of: class name "ActivityLoss"
input_types
output_types
__init__
forward 

It should be just like: 
https://github.com/NVIDIA-NeMo/Speech/blob/main/nemo/collections/asr/losses/bce_loss.py

File: nemo/collections/asr/losses/aux_diarization_loss.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        frame_losses = F.cross_entropy(
+            activity_logits.float().transpose(1, 2),
+            activity_targets,
+            reduction="none",
+        )
+        loss = (frame_losses * valid_frames).sum() / valid_frames.sum().clamp_min(1).float()
+    return loss
+
+
+def phantom_loss(
```
Comment (tango4j) [inline nemo/collections/asr/losses/aux_diarization_loss.py]: This is the same as activity_loss. 
loss should not be a function.
In NeMo Speech, it is expected to be a class that has forward method.

Please follow the convention of having:  class name "PhantomLoss"
input_types
output_types
__init__
forward 

It should be just like: 
https://github.com/NVIDIA-NeMo/Speech/blob/main/nemo/collections/asr/losses/bce_loss.py

**Review by tango4j — CHANGES_REQUESTED** [review]
Please make the changes in loss folder's aux_diarization_loss.py

File: examples/speaker_tasks/diarization/conf/neural_diarizer/sortformer_offline_8spk.yaml [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+    attn_mode: full  # Use full bidirectional self-attention.
+    drop_rate: 0.1  # Dropout probability in Transformer blocks.
+    dropout_pre_encoder: 0.1  # Dropout probability before the encoder.
+    dropout_emb: 0.0  # Input-embedding dropout probability.
+    sync_max_audio_length: true
+
+  loss:
+    _target_: nemo.collections.asr.losses.bce_loss.BCEWithLogitsLoss
+    reduction: mean
+
```
Comment (tango4j) [inline examples/speaker_tasks/diarization/conf/neural_diarizer/sortformer_offline_8spk.yaml:146]: here, the following should be added:

  activity_loss:
    _target_: nemo.collections.asr.losses.aux_diarization_loss.ActivityLoss
    reduction: mean

  phantom_loss:
    _target_: nemo.collections.asr.losses.aux_diarization_loss.PhantomLoss
    reduction: mean

File: examples/speaker_tasks/diarization/conf/neural_diarizer/sortformer_streaming_8spk.yaml [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    attn_mode: full  # Use full attention by default; selected training batches apply causal-tail masking.
+    drop_rate: 0.1  # Dropout probability in Transformer blocks.
+    dropout_pre_encoder: 0.1  # Dropout probability before the encoder.
+    dropout_emb: 0.0  # Input-embedding dropout probability.
+    sync_max_audio_length: true
+
+  loss:
+    _target_: nemo.collections.asr.losses.bce_loss.BCEWithLogitsLoss
+    reduction: mean
+
```
Comment (tango4j) [inline examples/speaker_tasks/diarization/conf/neural_diarizer/sortformer_streaming_8spk.yaml:169]: here, the following should be added:

activity_loss:
target: nemo.collections.asr.losses.aux_diarization_loss.ActivityLoss
reduction: mean

phantom_loss:
target: nemo.collections.asr.losses.aux_diarization_loss.PhantomLoss
reduction: mean

File: nemo/collections/asr/losses/aux_diarization_loss.py (hunk 8 above)
Comment (tango4j) [reply nemo/collections/asr/losses/aux_diarization_loss.py]: Loss class should be referred as follows: 

=========
phantom_loss:
target: nemo.collections.asr.losses.aux_diarization_loss.PhantomLoss
reduction: mean

File: nemo/collections/asr/losses/aux_diarization_loss.py (hunk 7 above)
Comment (tango4j) [reply nemo/collections/asr/losses/aux_diarization_loss.py]: Loss class should be referred as follows: 

=========
activity_loss:
target: nemo.collections.asr.losses.aux_diarization_loss.ActivityLoss
reduction: mean

**Review by tango4j — CHANGES_REQUESTED** [review]
Please construct the loss class as other loss files (just like bce_loss)

File: examples/speaker_tasks/diarization/neural_diarizer/e2e_diarize_speech.py (hunk 4 above)
Comment (ipmedenn) [reply examples/speaker_tasks/diarization/neural_diarizer/e2e_diarize_speech.py]: done

File: examples/speaker_tasks/diarization/neural_diarizer/checkpoint_averaging.py (hunk 3 above)
Comment (nithinraok) [reply examples/speaker_tasks/diarization/neural_diarizer/checkpoint_averaging.py:1]: Thanks, I learned this recently.

**Review by nithinraok — APPROVED** [review]
LGTM

**Review by tango4j — APPROVED** [review]
Changes are made exactly as requested.
